src/stdlib/runtime.c

```diff
@@ -393,25 +392,63 @@ void delete_value(value_p old_value) {
 	}
 }
 
-value_p getVal(struct var_instance *inst, int x, int y) {
-	if(!assertInBounds(inst, x, y)) return new_val();
-	int offset = inst->rows * y + x;
-	char *status = inst->status + offset;
-	value_p return_val;
-	if(*status & IN_PROGRESS) {
-		/* TODO: Circular dependency. Possibly throw? */
-		return_val = new_val();
-	} else if ((~(*status)) & CALCULATED) { /* value not calculated */
-		value_p val = calcVal(inst, x, y);
-		inst->values[offset] = val;
-		*status = (*status && !IN_PROGRESS) | CALCULATED;
-		return_val = val;
-	} else {
-		return_val = inst->values[offset];
-	}
-	while(return_val->flags == FLAG_SUBRANGE && return_val->subrange->subrangeRow == 1 && return_val->subrange->subrangeCol == 1) {
-		return_val = getVal(return_val->subrange->range, return_val->subrange->offsetRow, return_val->subrange->offsetCol);
+value_p getVal(struct var_instance *inst, int r, int c) {
+	/* If we're going to return new_val() then we have to
+	 * do clone_value(). Otherwise the receiver won't know
+	 * whether or not they can free the value_p they get back.
+	 * I think this should return, dangerously, return NULL if it's
+	 * invalid, and the callers will have to be careful to check the value.
+	 * The alternative is to always clone_value - safer, but much slower
+	 * and makes our memory issues even bigger.
+	 * Right now there are only a few places that call this. */
```

Comment:
We could add a null/empty object to each scope. That is returned in such situations.